README

NAME

    ETS -- Wolfenstein: Enemy Territory Statistics

AUTHOR

    (c) Vladi Shabanski <cade@bis.bg> 2003
    http://soul.datamax.bg/~cade/away/ets
    irc.freenode.net #ets
  
SYNOPSIS

    ets.pl
    ets-rt.pl
    ets-mb.pl
    ets-sw.pl

DESCRIPTION
  
    ETS reads Wolfenstein: Enemy Territory server logfile 
    and produces static web site with statistics. You can
    use it to produce several logfiles from different
    servers as long as you keep output directory different
    for each server/logfile.

    ETS-RT tracks server console log file and maintains
    run-time statistics (both in-game and web stats). It
    acts also as in-game bot and can perform various
    rcon commands.

    ETS-MB (master browser) gathers global statistics for all running servers.
    Most important information is popularity (player loads, maps, etc.).

    ETS-SW (servers watch) gathers online info from running servers.

    All tools described here require Perl (see REFERENCES below)

INSTALL

    Actually there is no real install procedure. There are 
    just few steps you should follow:
    
    1. extract ets-nnnnnn.tar.gz package
       copy ets.pl ets-rt.pl and ets_lib (directory!) to the
       location where you will run ETS from.
       
    1a. check ets_lib/html_inc dir if you want to edit html templates. 
       
    2. create file with environment config you need
       (see ENVIRONMENT CONFIG section in this document)
    
    3. create web-visible directory and copy recursively content of `themes'
       directory into it (not the themes-folder itself!).
    
    4a. start ets.pl without any arguments, i.e. something like:
    
       ets.pl
       
    That's all, usually ETS should be attached to crontab jobs but
    you can also run it manually.
    
    4b. if you are using environment config described below, you should do
        something like:
        
        # UNIX/bash:
        . ets_config
        ets.pl
        
        # DOS/Windows:
        ets_config.bat
        ets.pl

    5. if you want to use ets-rt.pl you have to create environment config
       (see ETS-RT ENVIRONMENT CONFIG section in this document)
       and then run 'ets-rt.pl' script. (note that ets-rt.pl will not exit
       as ets.pl but will continue to work until you stop it!)
       
    6. if you use ETS-RT for online web stats, you need to copy some screenshots
       from the game to `$OUTPUT_DIR/screenshots'. screenshots are the same
       that ET produces when you hit the screenshot key.

       ets package has some screenshots ready for this including special
       'no-screenshot.jpg' file which is shown when no screen is found for
       the current map. more screenshots can be found here:
       
            http://cade.datamax.bg/away/ets/ets-rt-screenshots-dir/
       
       NOTE: screenshots on the web page are limited to 400x300 pixels, so you
       better shrink your screens before put them on the web, so they will be
       smaller for users to download!
       
       More screenshots sized 400x300 with fixed gamma can be found here:
       
            http://cade.datamax.bg//away/ets/ets-rt-screenshots-dir/
            http://play.evrocom.net/games/et/et-stats/screenshots/


ENVIRONMENT CONFIG

    ETS is configured with environment variables.

    ETS_LIB=/path/to/ets_lib
    
        usually you don't have to set this one. ets.pl will look for it
        at the location ets.pl is installed. however if you need to move
        ets_lib to another location or automatic find doesn't work you'll
        need to set ETS_LIB manually!

    ETS_LOG_FILE=et_logfile
    
        this is source log file. this is required to be the "gamelog"
        it should look like this:
        
            (note: long lines cut)
            7:41 Item: 1 item_health
            7:41 Kill: 5 6 15: ^2USA-^5IceMan killed DOCTOR ROCKTER by MOD_FG42
            7:43 Kill: 3 5 43: ^-^2-=BC=-^1d3Gr3a73t killed ^2USA-^5IceMan by MO
            7:47 Kill: 1 4 9: ^(radko killed toko by MOD_MP40
            7:52 Item: 1 weapon_mp40
            8:13 Kill: 5 2 15: ^2USA-^5IceMan killed ^1Lg-^2aLieNs by MOD_FG42
            8:17 ClientUserinfoChanged: 3 n\^-^2-=BC=-^1d3Gr3a73t\t\1\c\2\r\1...
            8:18 sayteam: ^1Lg-^2aLieNs: are we Bv
            8:19 Kill: 1 4 9: ^(radko killed toko by MOD_MP40
            8:21 sayteam: ^1Lg-^2aLieNs: BC
            8:22 ClientUserinfoChanged: 7 n\^2[LUD]^1demented\t\3\c\0\r\0\m\...
            8:22 ClientBegin: 7
            8:26 sayteam: ^1Lg-^2aLieNs: samo za exp igrae6
            8:31 ClientBegin: 7
            8:43 WeaponStats: 4 0 2097240 0 0 0 2 0 2 33 0 0 0 8 56 0 0 ...
            8:43 ClientDisconnect: 4
            8:53 Kill: 0 2 10: Go6ko killed ^1Lg-^2aLieNs by MOD_THOMPSON
            8:53 Kill: 7 6 10: ^2[LUD]^1demented killed DOCTOR ROCKTER by MOD_TH
            8:58 ClientUserinfoChanged: 5 n\^2USA-^5IceMan\t\2\c\4\r\1\m\...
            9:00 Kill: 3 5 43: ^-^2-=BC=-^1d3Gr3a73t killed ^2USA-^5IceMan by MO
            9:00 ClientConnect: 4   
            9:00 Kill: 1 7 9: ^(radko killed ^2[LUD]^1demented by MOD_MP40
            9:01 Item: 1 item_health
            9:01 Item: 1 item_health

        NOTE: this is mandatory for ETS!
        
        to enable this log you need this option set in your server:
        
            g_log "games.log"

    ETS_CON_FILE=et_console_logfile
    
        this is source console log file. this is the "console log"
        it should look like this:
        
            (note: long lines cut)
            ^3PunkBuster Server: New Connection (slot #16) 192.168.107.112:27
            ^3PunkBuster Server: Player GUID Computed d1732c83453453452fcb3a3
            Userinfo: \cg_etVersion\ET Pro, ET 2.56\cg_uinfo\621 0 30\g_passw
            ClientUserinfoChanged: 15 n\^1RESIDENT ^0VALBO\t\3\c\0\r\0\m\0000
            voice: ^2The^0X^2Files^x*^1Mulder iamcovertops ^0-^1i^x*am*^2cove
            etpro IAC: 15 GUID [CEC5EA934534534534534534586833D4092947A0] [^1
            saybuddy: ^yShades*^fFinish him !!!: slagam din
            [lof](^yShades*^fFinish him !!!^7) (D,3): slagam din
            Item: 10 item_health
            Kill: 0 13 17: ETPlayer killed ^yShades*^fFinish him !!! by MOD_P
            Item: 9 weapon_magicammo
            Item: 10 item_health
            Item: 10 item_health
            saybuddy: ^yShades*^fFinish him !!!: ne
            [lof](^yShades*^fFinish him !!!^7) (D,3): ne
            Kill: 1 9 1: ^2[RIO]^1K^tOCbO killed ^h@dTa^0->^1Jason^5Din^1Alt 
            Kill: 7 4 9: ^0Hard^7Tastic killed ^1/S.W.A.T/^2C.Solo by MOD_MP4
            Kill: 13 1 10: ^yShades*^fFinish him !!! killed ^2[RIO]^1K^tOCbO 
            voice: ^yShades*^fFinish him !!! Medic
            Kill: 0 13 17: ETPlayer killed ^yShades*^fFinish him !!! by MOD_P
            Item: 7 item_health
            Setting Axis autospawn to West Bunker
            Setting Allied autospawn to Allied West Beach
            Client 35 connecting with 300 challenge ping
            Setting Axis autospawn to Axis Main Bunker
            Setting Allied autospawn to West Bunker
            Item: 7 item_health

        even though both logs seems the same, they hold different information!
        
        NOTE: this is optional for ETS but is mandatory for ETS-RT!


HISTORY
  ETS -- Wolfenstein: Enemy Territory Statistics
  (c) Vladi Shabanski <cade@bis.bg> 2003-2004
  http://cade.datamax.bg/away/ets
  
  HISTORY

  + feature added
  - feature removed or changed
  ! bugfix
  % note

  19.Jun.2005
  
    + ETS-SW (server watch) has been added. It produces page with online
      info for remote servers (which you do not own or have direct access).
    + Sponsor change: 
      Thanks to play.evrocom.net and furiozo for all this time! :) (2 years)
      Now ETS has own server kindly provided by:
      
            http://www.GameServerHost.nl
            
      Current address of ETS/LUD server is: 217.148.190.25:27971
      and it's status can be seen here: http://cade.datamax.bg/away/ets/sw/

  22.May.2005
  
    + game id's are now human readable time instead of unix time
    + player references for objectives table per game
    + fixed weapons tops accuracy ratings
    + more crossrefs for weapons usage
    ! integer figures equal to 0 are not show in tables (empty cells)
    ! top ranks link in menu finaly added :/ sorry tjw
    + top ranks table expanded with separate 'game' column
    ! noise player links fixed
    + frag spree duration
  
  02.Apr.2005
  
    + ETS-MB master browser and statistics has been added.

  30.Nov.2004
  
    + top player ranks added to each user's page
      (patch by tjw)

  31.Jul.2004

    ! now kills in the team are ignored.
    ! fixed ref_file, now accepts any case guids
    + ?who added '?who player' will show best known nick of this player

  26.Jul.2004
  
    + ets-red theme added
      inspired by http://et.gob.nu/stats/games.html
      thanks! :)
    + ets_lib added (large portion of ets.pl and ets-rt.pl cleaned up)
    + ets_lib/html_inc added. html templates.

  21.Jul.2004
  
    + ETS-RT real-time stats and in-game bot utility has been implemented.

  07.Jul.2004
  
    + jerks ladder 
      thanks tjw again :) here is his original page: http://et.tjw.org/gg_bg.php
      I also have stolen 'jerk' term from him :)
    + added ets-month-rotation.sh sample shell script which helps by month
      logs/stats rotation (backup)
      (tjw again, auto-thanks on:))  
  
  14.Jun.2004
    
    + `last games in which playeed...' list implemented for clan stats
    ! fixed clan recognitions
    + maps stats added: total games, allies/axis wins, xp wins theory
      thanx to tjw for the inspiration: http://et.tjw.org/teamwins_long.php
      :)
  
  27.May.2004 - orgoj007

    + some h2 tags chnaged to h3 and added class ets to all header tags
    + added class atribute to table tags
    + if player not found in defined clan map try default mapping
    + finalized localization and new czech localization
    + configurable header, footer and  color overrideing
    + cz007 theme added

  05.May.2004
  
    + frag spree (thanks tjw)
    + fixed transcription colors for ets-light theme
  
  12.Apr.2004
  
    + more config items (ETS_MIN_PLAYERS,ETS_FLOOR...)
    + ETPro 3.0 fixes
      thanx to orgoj@cz007.net
      
    + total K:D ratio on each player page
      thanx to Tony J. White (http://et.tjw.org) for this patch

  02.Apr.2004
  
    + harvest awards
    + sniper stats
    ! other fixups
  
  13.Jan.2004

    + top XP per single game score rank is back ;)
    + references for each game for every XP in the tops pages
    + objectives scores i.e. count of captures/repairs/dynamites/diffuses
    ! few small fixups

  04.Jan.2004
    
    + references for last games in which every player participated
    + which players were referees in the game transcription

  30.Nov.2003
  
    ! fixed ets-default theme.

ets_lib/ets.pm
##############################################################################
#
#  ETS -- Wolfenstein: Enemy Territory Statistics
#  (c) Vladi Shabanski <cade@bis.bg> 2003
#  http://soul.datamax.bg/~cade/away/ets
#
#  Developed for play.evrocom.net server ET community
#  http://play.evrocom.net
#
#  This software is distributed under GPL license, see COPYING for details.
#
#  $Id: ets.pm,v 1.8 2005/05/23 22:19:18 cade Exp $
#
##############################################################################
package ets;
use FindBin;
use lib "$FindBin::Bin/ets_lib";
use Exporter;
use strict;

our @ISA    = qw( Exporter );
our @EXPORT = qw(
                DEBUG
                
                $LOG_FILE     
                $LOG_CON_FILE 
                $LOG_PB_FILE  
                $OUTPUT_DIR   
                $WEB_PREFIX   
                $WEB_DIR      
                $THEME        
                $LAST_DAYS    
                $MOTD_FILE    
                $SERVER_URL   
                $MIN_PLAYERS  
                $FLOOR        
                $TOPS_FLOOR   
                $SCORE_MIN
                $XPWINS_DIFF  
                $XPJERK_DIFF  
                $WINS_MINXP   
                
                %CLASS
                %SIDE
                %WEAPON
                %WEAPON_REV
                %SKILL
                %SKILL_IMG
                %RANK
                %GAMETYPE
                %COLOR
                
                %WIN
                %ORIGINAL_MAP
                get_winner
                
                %NAM_COLOR
                $NAME_MAP
                $CLAN_MAP
                $TRANSLATION
                
                $REVISION
                $ETS_VER 
                $CREDITS
                
                get_clan
                
                %href
                mod_hash_stats
                
                hash_max
                hash_sum
                time_fmt

                make_columns
                html_table
                html_save
                color_name
                escape_name
                
                load_file save_file load_hash save_hash
                );

### CONFIG ###################################################################

our $ROOT = $ENV{ 'ETS_LIB' } || "$FindBin::Bin/ets_lib";

our $LOG_FILE      = $ENV{ 'ETS_LOG_FILE' };
our $LOG_CON_FILE  = $ENV{ 'ETS_CON_FILE' };
our $LOG_PB_FILE   = $ENV{ 'ETS_PB_FILE' };
our $OUTPUT_DIR    = $ENV{ 'ETS_OUTPUT_DIR' };
our $WEB_PREFIX    = $ENV{ 'ETS_WEB_PREFIX' };
our $WEB_DIR       = $ENV{ 'ETS_WEB_DIR' } || '.';
our $THEME         = $ENV{ 'ETS_THEME' } || "ets-default";
our $LAST_DAYS     = $ENV{ 'ETS_LAST_DAYS' };
our $MOTD_FILE     = $ENV{ 'ETS_MOTD_FILE' };
our $SERVER_URL    = $ENV{ 'ETS_SERVER_URL' } || 'http://play.evrocom.net'; # this is credit line, please preserve it!
our $MIN_PLAYERS   = $ENV{ 'ETS_MIN_PLAYERS' } || 4;
our $FLOOR         = $ENV{ 'ETS_FLOOR' } || 50;
our $TOPS_FLOOR    = $ENV{ 'ETS_TOPS_FLOOR' } || 15; # maximum player count on top scores page
our $SCORE_MIN     = $ENV{ 'ETS_SCORE_MIN' } || 100; # minimum player score to generate statistics
our $XPWINS_DIFF   = $ENV{ 'ETS_XPWINS_DIFF' } || 20; # minimum difference for xp wins theory calculation
our $XPJERK_DIFF   = $ENV{ 'ETS_XPJERK_DIFF' } || 30; # minimum difference for xp jerk ladder
our $WINS_MINXP    = $ENV{ 'ETS_WINS_MINXP' } || 200; # minimum xp to calculate wins
our $HTML_INC      = $ENV{ 'ETS_HTML_INC' }; # html include directory
our $CLAN_MAP_ONLY = $ENV{ 'ETS_CLAN_MAP_ONLY' }; # use only clan map, no clan recognition


our @HTML_INC = ( split( /\s+/, $HTML_INC ), "$ROOT/html_inc" );

### DEBUGGING ################################################################

our $DEBUG = $ENV{ 'DEBUG' };

sub DEBUG { print STDERR "DEBUG:@_\n" if $DEBUG; };

DEBUG( "START" );

### ET STATICS ###############################################################

our $REVISION = '$Revision: 1.8 $';
our $ETS_VER = $1 if $REVISION =~ /([\d\.]+)/;
our $CREDITS = "<small><a href=http://cade.datamax.bg/away/ets><b>ETS $ETS_VER</b></a> by (c) <a href=mailto:cade\@bis.bg>Vladi Belperchinov-Shabanski &lt;cade\@bis.bg&gt;</a> 2003</small>";

our %CLASS = (
             0 => 'SOLDIER',
             1 => 'MEDIC',
             2 => 'ENGINEER',
             3 => 'FIELD_OPS',
             4 => 'COVERT_OPS',
            );
            
our %SIDE  = (
             0 => 'UNKNOWN',
             1 => 'AXIS',
             2 => 'ALLIES',
             3 => 'SPECTATORS',
            );            

our %WEAPON = (
             0 => 'KNIFE',
             1 => 'LUGER',
             2 => 'COLT',             #  8
             3 => 'MP40',             #  9
             4 => 'THOMPSON',         # 10
             5 => 'STEN',             # 11
             6 => 'FG42',             # 15
             7 => 'PANZERFAUST',
             8 => 'FLAMETHROWER',
             9 => 'GRENADE',
            10 => 'MORTAR',
            11 => 'DYNAMITE',         # 26
            12 => 'AIRSTRIKE',        # 28
            13 => 'ARTY',             # 30
            14 => 'SYRINGE',
            15 => 'SMOKECAN',
            16 => 'SATCHEL',
            17 => 'GRENADE_LAUNCHER', # 18
            18 => 'LANDMINE',         # 45
            19 => 'MOBILE_MG42',
            20 => 'GARAND',
            21 => 'KAR98'
            );
our %WEAPON_REV = reverse %WEAPON;            

our %SKILL = (
             0 => 'BATTLE SENSE',             
             1 => 'ENGINEERING   (ENGINEER)',             
             2 => 'FIRST AID     (MEDIC)',
             3 => 'SIGNALS       (FIELD OPS)',
             4 => 'LIGHT WEAPONS',
             5 => 'HEAVY WEAPONS (SOLDIER)',
             6 => 'COVERT OPS    (COVERT OPS)',
            );

our %SKILL_IMG = (
             0 => "<img src=#bs.gif border=0>",
             1 => "<img src=#en.gif border=0>",
             2 => "<img src=#fa.gif border=0>",
             3 => "<img src=#si.gif border=0>",
             4 => "<img src=#lw.gif border=0>",
             5 => "<img src=#hw.gif border=0>",
             6 => "<img src=#co.gif border=0>",
            );

our %RANK = (
             0 => 'Private',
             1 => 'Private First Class',
             2 => 'Corporal',
             3 => 'Sergeant',
             4 => 'Lieutenant',
             5 => 'Captain',
             6 => 'Major',
             7 => 'Colonel',
             8 => 'Brigadier General',
             9 => 'Lieutenant General',
             10 => 'General'
            );

our %GAMETYPE = (
             2 => 'Single Map Objective',
             3 => 'Stopwatch',
             4 => 'Campaign',
             5 => 'Last Man Standing',
            );
            
# thanx, furiozo :)
our %COLOR = (
            '1' => 'f70400',
            '2' => '02f702',
            '3' => 'fdf900',
            '4' => '0006f8',
            '5' => '00f6f3',
            '6' => 'f407f5',
            '7' => 'f9f9f9',
            '8' => 'ff7b06',
            '9' => '7c7d81',
            '0' => '000100',
            'q' => 'f70303',
            ')' => 'f70303',
            'w' => 'f4f6f1',
            'e' => '8100fd',
            'r' => '00f904',
            't' => '0005f4',
            'y' => '7b7d7a',
            'u' => '00faf6',
            'i' => 'fb0033',
            'o' => 'eff174',
            'p' => '060702',
            'a' => 'fd9534',
            's' => 'fbf904',
            'd' => '007dfe',
            'f' => '3795c8',
            '&' => '3795c8',
            'g' => 'c2fcc9',
            'h' => '066432',
            '(' => '066432',
            'j' => 'ad1c17',
            'k' => '9b3107',
            'l' => 'cf9834',
            ';' => 'c0bfc7',
            'z' => 'bcbdc2',
            'x' => 'f98000',
            'c' => '7b0484',
            'v' => 'f900f9',
            'b' => '037d7e',
            'n' => 'faf9c0',
            'm' => '9b9935',
            ',' => 'cf9a3a',
            '.' => 'f7f8b6',
            '/' => 'fbfc7c',
            '-' => '98972d',
            '=' => '7f7f00',
            '\\' => '007f03',
            '[' => 'bec0bb',
            ']' => '7f7c0b',
            '\'' => 'c5f9c8',
            '*' => 'ae181a',
            );

# color map customization
%COLOR = ( %COLOR, load_hash( $ENV{ 'ETS_COLOR_OVERRIDE' } ) );

### WIN/LOSE MAP #############################################################

# grep InitGame | perl -ne 'print "$1\n" if /mapname\\([^\\]+)/' | sort | uniq

# this map tells who is winner if time limit is reached, i.e. who defends the map

our %WIN = (
              '2hide'              => 'AXIS',
              'tc_base'            => 'AXIS',
              'apennines_b2'       => 'AXIS',
              'atlantic'           => 'AXIS',
              'base'               => '',
              'battery'            => 'AXIS',
              'bzz_drunk'          => '',
              'carnage2'           => 'AXIS',
              'etsbay'             => 'AXIS',
              'flughafen'          => '',
              'fueldump'           => 'AXIS',
              'goldrush'           => 'AXIS',
              'maiden_beta4'       => '',
              'mml_church_et_v1'   => '',
              'mp_sillyctf'        => '',
              'nightcrawlers'      => 'AXIS',
              'normandy_final'     => '',
              'oasis'              => 'AXIS',
              'venice'             => 'AXIS',
              'password2'          => 'AXIS',
              'radar'              => 'AXIS',
              'raiders_b2'         => 'AXIS',
              'railgun'            => 'ALLIES',
              'rhine'              => 'AXIS',
              'rhine2'             => 'AXIS',
              'rhine_bridge'       => 'AXIS',
              'supplydepot'        => 'AXIS',
              'supplydepot2'       => 'AXIS',
              'temple'             => 'AXIS',
              'temple2'            => 'AXIS',
              'temple3'            => 'AXIS',
              'temple_final'       => 'AXIS',
              'stalingrad'         => 'AXIS',
              'fun_beach'          => 'AXIS',
              'fun_beach_final'    => 'AXIS',
              'trainyard'          => '',
              'v2_base_fp'         => 'ALLIES',
              'v2base'             => 'ALLIES',
              'caen'               => 'ALLIES',
              '1944_beach'         => 'AXIS',
              'axislab_final'      => 'AXIS',
              'ammodepot'          => 'AXIS',
          );

our %ORIGINAL_MAP = (
              'battery'            => 1,
              'fueldump'           => 1,
              'goldrush'           => 1,
              'radar'              => 1,
              'railgun'            => 1,
              'oasis'              => 1,
          );

sub get_winner
{
  my $map = lc shift();
  my $tlh = shift() ? 1 : 0;
  return undef unless $WIN{ $map };
  return $WIN{ $map } if $tlh;
  return $WIN{ $map } eq 'AXIS' ? 'ALLIES' : 'AXIS';
}

### VARS ETC #################################################################

our %NAM_COLOR;
our $NAME_MAP;
our $CLAN_MAP;    
our $TRANSLATION; 

our %href;

### TABLE FORMATTING #########################################################

sub make_columns
{
  my $text;
  
  my $c = @_;
  my $p = $c > 0 ? int( 100 / @_ ) : 100;
  
  $text .= "<table cellspacing=5 width=100% class=etscolumns><tr><td width=p% valign=top>";
  $text .= join "</td><td width=$p% valign=top>", @_;
  $text .= "</td></tr></table>";
  
  return $text;
}

sub html_table_cmp
{
  my $aa = shift;
  my $bb = shift;
  my $ta = shift; # row type
  my $tb = shift; # row type
  my $sa = shift; # sort alpha
  
  return -1 if $ta eq 'HEAD';
  return  1 if $tb eq 'HEAD';
  return  1 if $ta eq 'FOOT';
  return -1 if $tb eq 'FOOT';
  
  $aa = $3 if $aa =~ /^(\[([^\]]+)\])?(.+)(\/([ip]))?$/o;
  $bb = $3 if $bb =~ /^(\[([^\]]+)\])?(.+)(\/([ip]))?$/o;
  
  if( ! $sa )
    {
    $aa = $1 if $aa =~ /^([\+\-\d\.]+)/;
    $bb = $1 if $bb =~ /^([\+\-\d\.]+)/;
    }
  
  return $sa ? $aa cmp $bb : $bb <=> $aa;
}

sub html_table
{
  my $ar = shift;
  my %opt = @_;
  
  my $text;
  $text .= "<table class=ets width=100% cellspacing=1 cellpadding=5>\n";
  my @or; # options row
  my @ft; # format type
  my $r = -1; # row
  
  my @ar = @$ar;
  
  my $floor = $opt{ 'FLOOR' } || $FLOOR;
  
  my $oc = $opt{ 'SORT' } || $opt{ 'SORTA' }; # sort/order column
  @ar = sort { html_table_cmp( $a->[$oc], $b->[$oc], $a->[0], $b->[0], $opt{ 'SORTA' } ) } @ar if $oc > 0;
  
  for my $rr ( @ar )
    {
    $r++ if $rr->[0] eq '';
    if( $floor > 0 )
      {
      next if $rr->[0] eq 'COMT' and $r >= $floor;
      next if $rr->[0] eq ''     and $r >= $floor;
      }
    my $c = -1; # column
    my $class = $r % 2 ? 'r1' : 'r2';
    @or = @$rr and next if $rr->[0] eq 'OPT';
    $class = 'hh' if $rr->[0] eq 'HEAD'; # header
    $class = 'ff' if $rr->[0] eq 'FOOT'; # footer
    $class = 'cc' if $rr->[0] eq 'COMT'; # comments
    $text .= "<tr class=$class>\n";
    for( @$rr )
      {
      my $d = $_;
      $c++;
      next if $c == 0;
      
      if( $d =~ s/^\[([^\]]+)\]//o )
        {
        my $p = $1; # params
        my $o; # opts
        $o .= " align=right" if $p =~ />/o;
        $o .= " width=$1"    if $p =~ /(\d+%)/o;
        $or[$c] = $o;
        $ft[$c] = lc $1 if $p =~ /([ip])$/io;
        #print STDERR "TABLE OPT: $p --> $o\n";
        #$text .= "TABLE OPT: $p --> $o<br>";
        }
      my $f;
      $f = $ft[$c] if $rr->[0] ne 'HEAD' and $rr->[0] ne 'FOOT';
      $f = lc $1 if $d =~ s/\/([ip])$//io; # [i]nt, [p]ercent
      $d = table_sum( $ar, $c ) if $d eq '(SUM)';
      $d = table_ave( $ar, $c ) if $d eq '(AVE)';
      $d = table_med( $ar, $c ) if $d eq '(MED)';

      my $cmt;
      if( $f eq 'i' or $f eq 'p' )
        {
        ( $d, $cmt ) = split /\|/, $d, 2;
        }
      
      $d = num_fmt( int $d ) if $f eq 'i';
      $d = $d > 0 ? sprintf "%.2f%%", $d : '' if $f eq 'p';

      $d = '' if $f eq 'i' and $d == 0;
      
      $d = $r + 1 if $d eq '(ROW)';
      $d = "<span class=hitext>$d</div>" if $oc > 1 and $c == $oc and ! $$rr[0];
      $text .= "<td class=etstd $or[$c]>$d$cmt</td>\n";
      }
    $text .= "</tr>\n";
    }
  $text .= "</table>\n";
  return $text;
}

sub table_sum
{
  my $ar = shift;
  my $col = shift;
  my $sum = 0;
  for my $rr ( @$ar )
    {
    next if $rr->[0] ne ''; # skip HEAD,FOOT
    $sum += $rr->[$col];
    }
  return $sum;
}

sub table_ave
{
  my $ar = shift;
  my $col = shift;
  my $sum = 0;
  my $cnt = 0;
  for my $rr ( @$ar )
    {
    next if $rr->[0] ne ''; # skip HEAD,FOOT
    $sum += $rr->[$col];
    $cnt++;
    }
  return $cnt > 0 ? $sum / $cnt : 0;
}

sub table_med
{
  my $ar = shift;
  my $col = shift;
  my @sum;
  for my $rr ( @$ar )
    {
    next if $rr->[0] ne ''; # skip HEAD,FOOT
    push @sum, $rr->[$col];
    }
  @sum = sort { $b <=> $a } @sum;  
  return $sum[ @sum / 2 ];
}

### FORMATTING ###############################################################

sub time_fmt
{
  my $data = shift;
  my $d = int($data / (24*60*60));
  my $h = int(($data % (24*60*60)) / (60*60));
  my $m = int(($data % (60*60)) / 60);        
  my $s = $data % 60;                 
  return sprintf "%6dd %02dh %02dm %02ds", $d, $h, $m, $s;
}

sub num_fmt
{
  my $s = shift;
  $s = reverse $s;
  1 while $s =~ s/(\d\d\d)(\d)([^']*)$/$1'$2$3/;
  $s = reverse $s;
}

### ETS ######################################################################

sub color_name
{
  my $s = shift;
  my $cs = $NAM_COLOR{ $s } || $s;
  # return $s unless $cs;
  my $c = $cs =~ s/\^(.)/"<font color=#" . $COLOR{ lc $1 } . ">"/gie;
  $cs .= "</font>" x $c;
  return "<span class=name>$cs</name>";
}

sub hash_max
{
  my $hr = shift;
  my $max_k = undef;
  my $max_v = undef;
  while( my ( $k, $v ) = each %$hr )
    {
    if( $max_v eq '' or $v > $max_v )
      {
      $max_k = $k;
      $max_v = $v;
      next;
      }
    }
  #print STDERR "$max_k, $max_v\n";
  return ( $max_k, $max_v );
}

sub hash_sum
{
  my $hr = shift;
  my $sum = 0;
  while( my ( $k, $v ) = each %$hr )
    {
    next if $k =~ /^!/;
    $sum += $v;
    }
  return $sum;
}

our %CLAN_CACHE;
sub get_clan
{
  my $s = shift;
  my $cc = $CLAN_CACHE{ $s };
  return $cc if $cc;
  return uc '!' . $CLAN_MAP->{ $s } if $CLAN_MAP->{ $s };
  return '!' if $CLAN_MAP_ONLY;
  my $c;
  $c = $1 if ! $c and $s =~ /^\[([^\]]+)\]/o;
  $c = $1 if ! $c and $s =~ /\[([^\]]+)\]$/o;
  $c = $1 if ! $c and $s =~ /^([A-Z]{2,6})[:\-\|]/io;
  $c = $1 if ! $c and $s =~ /^=(.+?)=/o;
  $c = $1 if ! $c and $s =~ /^-(.+?)-/o;
  $c = $1 if ! $c and $s =~ /^=\|(.+?)\|=/o;
  $c = $1 if ! $c and $s =~ /^-\|(.+?)\|-/o;
  $c = $1 if ! $c and $s =~ /^<(.+?)>/o;
  return $CLAN_CACHE{ $s } = uc "!$c";
}

sub escape_name
{
  my $s = shift; 
  $s =~ s/([^a-zA-Z0-9\-\_])/sprintf("%02X", ord($1))/ge;
  return $s;
}

### MODS #####################################################################

#----------------------------------------------------------------------------

sub mod_hash_stats
{
  my $title = shift;
  my $hr    = shift;
  my %opt   = @_;
  my $text;

  my $clan = $opt{ 'CLAN' };
  my $tfmt = $opt{ 'TIMEFMT' };
  my $head = $opt{ 'HEAD' };
  my $href_prefix = $opt{ 'HREF_PREFIX' } || 'p_';
  
  my $sum = hash_sum( $hr );
  
  my @data;
  
  $text .= "<p>";
  $text .= "<h3 class='ets'>$title</h3>";
  
  push @data, [ 'OPT', "width=05% align=right", "width=40%", "width=30% align=right", "width=25% align=right" ];
  if( $head )
    {
    push @data, [ 'HEAD', split /\s*,\s*/, $head ];
    }
  else
    {  
    push @data, [ 'HEAD', "[>5%]#", "[<55%][T:Name]", "[>20%][T:Value]", "[>20%][T:Percentage]" ];
    }
  
  my $row = 0;
  for my $k ( sort { $hr->{ $b } <=> $hr->{ $a } } keys %$hr )
    {
    next if $clan and $k !~ /^!/;
    next if !$clan and $k =~ /^!/;
    
    my $v = $hr->{ $k };
    my $p = $sum > 0 ? sprintf "%.2f%%", 100 * $v / $sum : 'n/a'; # '
    
    my $k_colors = $opt{ 'NO_COLORS' } ? $k : color_name( $k );
    if( $href{ $k } )
      {
      $k = "<a class=t href=!$href_prefix$href{$k}.html>$k_colors</a>";
      }
    else
      {
      $k = $k_colors;
      }  
    
    $p = 'n/a' if $opt{ 'NOP' };
      
    if ( $row < $FLOOR )
      {
      $row++;
      $v = time_fmt( $v ) if $tfmt;
      push @data, [ '', '(ROW)', $k, $v, $p ];
      }        
    }
  $sum = time_fmt( $sum ) if $tfmt;
  if( $opt{ 'NOP' } )
    {
    pop( @$_ ) for ( @data );
    }
  else
    {  
    push @data, [ 'FOOT', '', '[>][T:Total]:', $sum, '' ];
    }
  $text .= html_table( \@data, FLOOR => $opt{ FLOOR } );
  $text .= "\n\n\n\n";
  return $text;
}

### IO #######################################################################

sub html_inc_file
{
  my $name = lc shift;
  my $html;
  for( @HTML_INC )
    {
    my $f = "$_/$name.html";
    next unless -e $f;
    #DEBUG( "html_inc_file: $f" );
    $html = load_file( $f );
    last;
    }
  $html =~ s/\[\#(\S+)\]/html_inc_file( $1 )/ge;
  return $html;
}

sub html_save
{
  my $file = shift;
  $file = "$OUTPUT_DIR/$file.html";
  
  my %DATA = %ENV;
  
  $DATA{ 'CREDITS' } = $CREDITS;
  $DATA{ 'FLOOR' } = $FLOOR;
  $DATA{ 'THEME' } = $THEME;
  $DATA{ 'TEXT' } = join '', @_;
  $DATA{ 'TIME' } = scalar localtime();
  $DATA{ 'LAST_DAYS' } = "Note: Statistics data for last $LAST_DAYS days only is used.<br>";
  
  my $html = html_inc_file( 'index' );
  
  $html =~ s/\[\$(\S+)\]/$DATA{ $1 }/g;
  $html =~ s/href=!/href=$WEB_PREFIX/gio;
  $html =~ s/href=#/href=$WEB_DIR/gio;
  $html =~ s/src=#/src=$WEB_DIR\/$THEME\/img\//gio;
  $html =~ s/\[T\:([^\]]*)\]/$TRANSLATION->{ $1 }||$1/gie; # \: is to avoid get_trans.pl
  save_file( $file, $html ) or die "$0: cannot save html file: $file ($!)\n";
  #print STDERR "file $file written ok\n";
}

sub load_file
{
  my $file = shift;
  open( my $i, $file ) or return undef;
  local $/ = undef;
  my $s = <$i>;
  close $i;
  return $s;
}

sub save_file
{
  my $file = shift;
  open( my $o, ">$file" ) or return undef;
  print $o @_;
  close( $o );
  return 1;
}

sub load_hash
{
  my $file = shift;
  my %opt = @_;
  my %h;
  for( split( /[\n\r]+/, load_file( $file ) ) )
    {
    next unless /(.+?)(?<!\\)=(.*)/;
    my $k = $1;
    my $v = $2;
    $k =~ s/\\(.)/$1/go;
    $v =~ s/\\(.)/$1/go;
    $k = uc $k if $opt{ 'KEY_UC' };
    $k = lc $k if $opt{ 'KEY_LC' };
    $v = uc $v if $opt{ 'VAL_UC' };
    $v = lc $v if $opt{ 'VAL_LC' };
    $h{ $k } = $v;
    }
  return \%h;  
}

sub save_hash
{
  my $file = shift;
  my $hr = shift;
  open( my $o, ">$file" ) or return undef;
  while( my ( $k, $v ) = each %$hr )
    {
    $k =~ s/=/\\=/g;
    print $o "$k=$v\n";
    }
  close( $o );
  return 1;
}

##############################################################################
1;
### EOF ######################################################################

